### PR TITLE
Add jsclient to simplewallet

### PR DESCRIPTION
**Pull request by prp20**
Co-authored-by: Ashish K Mishra<ashish.k.mishra@intel.com>

Signed-off-by: Prasad P R <ppprasad112@gmail.com>

### CODE REVIEW

File: jsclient/app.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+// BodyParser Middleware
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({ extended: false }));
+
+// Set Static Folder
+app.use(express.static(path.join(__dirname, 'public')));
+
+app.use('/', routes);
+//app.use('/users', users);
```
Comment (askmish) [inline jsclient/app.js]: Could you remove this redundant comment.

File: jsclient/public/js/main.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    sessionStorage.clear();
+    window.location.href = "/login";
+    alert("Successfully Logged out");
+}
+
+//Function to deposit money to the specified user account
+function depositMoney() {
+    var userDetails = sessionStorage.getItem('userId');
+    var amount = document.getElementById("depositAmt").value;
+    if (amount == 0 || amount == " ") {
```
Comment (askmish) [inline jsclient/public/js/main.js]: amount should be verified with empty string rather than a space?

Comment (askmish) [reply jsclient/public/js/main.js]: This comment is applicable to other parts of the code.

File: jsclient/public/js/main.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+//function to implement transfer function form client side
+function transferMoney() {
+    var userDetails = sessionStorage.getItem('userId');
+    var beneficiary = document.getElementById('beneficiaryUserId').value;
+    var amount = document.getElementById("transferAmt").value;
+    if (amount == 0 || amount == " ") {
+        alert("Please enter amount");
+    }else if(beneficiary == ""){
```
Comment (askmish) [inline jsclient/public/js/main.js]: This should be just another if condition, instead of else if.

File: jsclient/routes/SimpleWalletClient.js [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+FAMILY_NAME='simplewallet'
+
+function hash(v) {
+    return createHash('sha512').update(v).digest('hex');
+}
+
+class SimpleWalletClient {
+    constructor(userid) {
+	const privateKeyStrBuf = this.getUserPriKey(userid);
+        const privateKeyStr = privateKeyStrBuf.toString().trim();
```
Comment (askmish) [inline jsclient/routes/SimpleWalletClient.js]: Extra space in indentation.

File: jsclient/routes/SimpleWalletClient.js [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+	const context = createContext('secp256k1');
+	const privateKey = Secp256k1PrivateKey.fromHex(privateKeyStr);
+	this.signer = new CryptoFactory(context).newSigner(privateKey);
+	this.publicKey = this.signer.getPublicKey().asHex();
+        this.address = hash("simplewallet").substr(0, 6) + hash(this.publicKey).substr(0, 64);
```
Comment (askmish) [inline jsclient/routes/SimpleWalletClient.js]: Exra space in indentation.

File: jsclient/routes/SimpleWalletClient.js [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        console.log(userid);
+        console.log("Current working directory is: " + process.cwd());
+	var userprivkeyfile = '/root/.sawtooth/keys/'+userid+'.priv';
+        return fs.readFileSync(userprivkeyfile);
+    }
+
+    getUserPubKey(userid) {
+        console.log(userid);
+        console.log("Current working directory is: " + process.cwd());
+	var userpubkeyfile = '/root/.sawtooth/keys/'+userid+'.pub';
```
Comment (askmish) [inline jsclient/routes/SimpleWalletClient.js]: Indentation.

File: jsclient/routes/SimpleWalletClient.js [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+	console.log("wrapping for: " + this.address);
+	var inputAddressList = [address];
+	var outputAddressList = [address];
+       if (action === "transfer") {
+	   const pubKeyStrBuf = this.getUserPubKey(values[1]);
+           const pubKeyStr = pubKeyStrBuf.toString().trim();
+           var toAddress = hash("simplewallet").substr(0, 6) + hash(pubKeyStr).substr(0, 64);
+           inputAddressList.push(toAddress);
+           outputAddressList.push(toAddress);
+
```
Comment (askmish) [inline jsclient/routes/SimpleWalletClient.js]: Indentation and remove empty line.

File: jsclient/routes/SimpleWalletClient.js [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+	   payload = action+","+values[0]+","+pubKeyStr;
+       } else {
+	   payload = action+","+values[0];
+       }
+
+	var enc = new TextEncoder('utf8');
+	const payloadBytes = enc.encode(payload);
+	console.log("sending = " + payloadBytes);
```
Comment (askmish) [inline jsclient/routes/SimpleWalletClient.js]: This console log is not required. Please remove.

File: jsclient/routes/SimpleWalletClient.js [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+            	signerPublicKey: this.signer.getPublicKey().asHex(),
+            	transactionIds: transactions.map((txn) => txn.headerSignature),
+	        }).finish();
+
+		const batchSignature = this.signer.sign(batchHeaderBytes);
+	        const batch = protobuf.Batch.create({
+                header: batchHeaderBytes,
+                headerSignature: batchSignature,
+                transactions: transactions,
+       		 });
```
Comment (askmish) [inline jsclient/routes/SimpleWalletClient.js]: Indentation.

File: jsclient/routes/SimpleWalletClient.js [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+		const batchListBytes = protobuf.BatchList.encode({
+            	batches: [batch]
```
Comment (askmish) [inline jsclient/routes/SimpleWalletClient.js]: Indentation.

File: jsclient/routes/SimpleWalletClient.js [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+        	}).finish();
+	
+		this._send_to_rest_api(batchListBytes);	
+	}
+
+	_send_to_rest_api(batchListBytes){
+
```
Comment (askmish) [inline jsclient/routes/SimpleWalletClient.js]: Empty line.

File: jsclient/routes/SimpleWalletClient.js [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+		if (batchListBytes == null)
+		{
+                    var geturl = 'http://rest-api:8008/state/'+this.address
```
Comment (askmish) [inline jsclient/routes/SimpleWalletClient.js]: Indentation.

File: jsclient/routes/SimpleWalletClient.js [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+                        var amount = new Buffer(data, 'base64').toString();
+                        return amount;
+		   })
+		   .catch((error) => {
+ 		   console.error(error);
+		  }); 	
+		}
+		else{
+		    fetch('http://rest-api:8008/batches', {
+ 		    method: 'POST',
```
Comment (askmish) [inline jsclient/routes/SimpleWalletClient.js]: Indentation.

File: jsclient/routes/index.js [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    var amount = req.body.money;
+    var client = new SimpleWalletClient(userId);
+    client.transfer(beneficiary, amount);    
+    res.send({ message:"Amount "+ amount +" successfully added to " + beneficiary});
+});
+
+router.post('/balance', function(req, res){
+    var userId = req.body.userId;
+    var client = new SimpleWalletClient(userId);
+    var x = client.balance();
```
Comment (askmish) [inline jsclient/routes/index.js]: A better variable name instead of just 'x'.

**Review by askmish — CHANGES_REQUESTED** [review]

**Comment by danintel** [comment]
Since you're adding a JavaScript client, you need to add JavaScript to the list of supported clients in README.md under `Components`  and `Contributing`.

**Review by danintel — CHANGES_REQUESTED** [review]
Since you're adding a JavaScript client, you need to add JavaScript to the list of supported clients in README.md under Components and Contributing.

**Review by danintel — APPROVED** [review]

**Review by askmish — APPROVED** [review]

**Pull request merged**